### ISSUE TITLE
Migrate Zed to objc2

### ISSUE THREAD

**Issue by osiewicz** [zed-industries/zed#22408]
### Check for existing issues

- [X] Completed

### Describe the feature

It seems like the ecoystem is moving towards using [objc2](https://github.com/madsmtm/objc2/) instead of objc as the latter is unmaintained. Doing so would probably allow us to drop media crate altogether (plus it'd be a win in terms of interacting with Mac platform code).
Related: https://github.com/zed-industries/zed/issues/17467

### Zed Version and System Specs

Zed: v0.168.0 (Zed Nightly 564936e1fee313efcd4643c5e78472f6dc44ffc5)
OS: macOS 15.1.0
Memory: 64 GiB
Architecture: aarch64

**Comment by kvark** [comment]
I've recently completed moving Blade to objc2, since it was blocking my blade-graphics-0.6 release. Seems to be working well: there are zero instances of `msg_send!` left in code now!

Might be worth reconsidering for Zed to just switch to Blade on macOS and drop the metal-rs based graphics backend, to save time instead of porting it. I realize that it may require a few changes in the presentation, related to how it may need to be done in CoreAnimation transaction in some cases but not the others. Happy to help get them in!

The Zed on Blade Metal path is currently available via "gpui/macos-blade" feature, and the latest Blade uprev is coming in #22409.

**Comment by tidely** [comment]
With #30347 merged, we're now up to date with the latest objc2 when it comes to dependencies. If Zed is interested in this kind of work, I'd be willing to start working on moving from `objc` to the `objc2-*` crate family. If yes, it could be a good idea to create a TODO list or a roadmap of sorts for the migration.

**Comment by stormasm** [comment]
I asked a question about this subject on [discord](https://discord.com/channels/869392257814519848/1199799855007158352/1414011431833309225) but got no official response from anyone on the Zed team but some other contributors and I had a brief conversation about this subject so I thought it made more sense to file it here as part of this issue to keep track of whats going to happen moving forward.

gpui uses the [metal crate](https://github.com/gfx-rs/metal-rs) which is deprecated.

There is a `TODO` note in [gpui Cargo.toml](https://github.com/zed-industries/zed/blob/main/crates/gpui/Cargo.toml) to replace the metal crate with objc2

```rust
objc2 = { version = "0.6", optional = true }
objc2-metal = { version = "0.3", optional = true }
#TODO: replace with "objc2"
metal.workspace = true
```

Blade currently uses objc2 and objc2-metal in the [Blade Renderer](https://github.com/zed-industries/zed/blob/main/crates/gpui/src/platform/blade/blade_renderer.rs)

Does it make sense to have this be on the roadmap prior to publishing out `gpui` to [crates.io](https://crates.io/search?q=gpui) ?

**Comment by stormasm** [comment]
This is fairly obvious from the above comments but Zed also has references to the cocoa crate which is also deprecated...

- https://crates.io/crates/cocoa
- https://docs.rs/cocoa/0.26.1/cocoa/
- https://github.com/servo/core-foundation-rs/pull/731

This is kind of going to be a show stopper when Zed does eventually publish out `gpui` as a stand alone crate

**Comment by osiewicz** [comment]
On one hand it would be great to land this change incrementally, but on the other there's a fundamental incompatibility between `objc2` and `objc` types; hence, we'll probably have to do it in one go.. which is pretty time-consuming, error-prone and does not give us immediate returns. I started hacking on this last week, but I don't have high hopes of pushing it over the line in few days or whatever.

**Comment by tidely** [comment]
Yeah the medal renderer is likely something which will need to be migrated in one go. I’ve mostly been looking at the platform code which also used objc for almost everything.

I also took a go at moving the platform code to objc2 it a while back, there’s surprisingly much which can be done incrementally.

There are quite a few standalone functions which can be migrated independantly. There’s also a common pattern of accessing a macos global and changing a few settings, which yes will create functions which mix objc2 and objc (yuck) but might still make migration easier.

There are however some large custom gpui classes which will undeniably need more work and big PR’s aswell.

Considering we already pull in objc2 through blade I don’t think it’s far fetched to think we could take a shot at it. A lot of the API’s provided by objc2 are marked safe and are less error prone to work with than objc, which will certainly help with migration, but since platform and rendering code is such a critical component and testing being difficult, I can see why this needs to be done carefully.

The main benefit I see behind this is removing a considerable chunk of Zed’s unsafe code, which is always good.

**Comment by tidely** [comment]
Building Zed with the latest Rust version `1.96` gives out a warning for the [block](https://crates.io/crates/block) crate.

```
warning: the following packages contain code that will be rejected by a future version of Rust: block v0.1.6
```

More specifically:

```
> warning: static of uninhabited type
>   --> /Users/<username>/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/block-0.1.6/src/lib.rs:64:5
>    |
> 64 |     static _NSConcreteStackBlock: Class;
>    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>    |
>    = note: uninhabited statics cannot be initialized, and any access would be an immediate error
>    = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
>    = note: for more information, see issue #74840 <https://github.com/rust-lang/rust/issues/74840>
```

The `block` crate hasn't been updated in a decade, so I assume the two options are vendoring the crate, or migrating to the `block2` crate which is under the `objc2` family of crates. The latter would likely require more extensive migration to the objc2 ecosystem because of the usage of concrete types inside of the `metal` crate.

I think the old `objc` crate ecosystem having crates not become compilable seems like a good reason to explore a possible migration. I've expressed previously that migration may be surprisingly incremental. I'd love to hear your thoughts on this.

**Comment by Vanuan** [comment]
In the mean time, fixing up cocoa versions to non-deprecated ones will help to avoid GPUI build failures with strict `Dwarnings` flag. Note that pinning down just cocoa isn't enough:
```
cocoa-foundation v0.2.1
└── cocoa v0.26.0
    ├── gpui_apple v0.1.0 (~/zed/crates/gpui_apple)
    │   └── gpui_macos v0.1.0 (~/zed/crates/gpui_macos)
    │       └── gpui_platform v0.1.0 (~/zed/crates/gpui_platform)
```
v0.2.1 is where deprecation warnings are introduced:
https://github.com/servo/core-foundation-rs/pull/732#issuecomment-5461528876

**Comment by Vanuan** [comment]
I don't have mac, but looks like the objc2 migration work trivially happens in forks:
https://github.com/freefcw/fc-gpui/commit/7146771a14a6687b81566fe4ca844287d227ee72
https://github.com/freefcw/fc-gpui/commit/53aa801d013f973b6153dcb7068307b37bfbdef4
https://github.com/freefcw/fc-gpui/commit/76d19f8540ab5b96c566994fd2bb7d2f934b3847
so I don't see how this would be complex in upstream.

I've submitted a little fix PR that fixes a trivial build failure, but it looks like getting rid of cocoa dependency is a much more involved PR:
https://github.com/zed-industries/zed/pull/63378
